### ISSUE TITLE
Filter language is inconsistent across Haskell/Python

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#1010]
Originally reported of Google Code with ID 954.

```
I tried to figure out how to use `gnt-job list --filter`. My first try:

   # gnt-job list --filter '["=", "reason", "myreason"]'
   Error while parsing query filter: Failed to parse query filter '["=", "reason", "myreason"]': Expected "not" (at char 0), (line:1, col:1)
   ["=", "reason", "myreason"]
   ^
   Expected "not" (at char 0), (line:1, col:1)

`expected "not"` - that's probably not intended.

Further, in `qlang.py` `BuildFilterParser`, the `binopstbl` is inconsistent in a few ways:

- the constants are hardcoded, not using `_constants.py`
- OP_EQUAL is `==` while its Haskell counterpart is `=` (man 7 ganeti specifies `==`)
- OP_NOT_EQUAL does not exist in Haskell nor is it translated to `not(equal)`, using it fails
```
Originally added on 2014-09-23 16:32:49 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Example for when `!=` fails:

    # gnt-job list --filter 'status != "canceled"'
    Unhandled protocol error while talking to the master daemon:
    Failed to parse request: Cannot convert value '["job", ["id", "status", "summary"], ["!=", "status", "canceled"]]', error: Unknown filter operand '!='

```
Originally added on 2014-09-23 16:39:26 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
In addition, the `not` operator doesn't seem to work:

   gnt-job list

gives me a list of jobs, some of which are `cancelled`, some of which have `success`, and

   gnt-job list --filter 'status == "canceled"'

shows the canceled ones but

   gnt-job list --filter 'not (status == "canceled")'

doesn't show the `success` ones (it matches 0 jobs).
```
Originally added on 2014-09-23 16:42:58 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-09-24 10:57:15 +0000 UTC.

Added Labels: Type-Defect Priority-High Component-UI Usability
Added to Milestone: Release2.13

**Comment by Ganeti-Issues-Migrator** [comment]
```
Investigating further, it seems that only `gnt-job list` is affected; for `gnt-node list`, the `not` filter works fine.

It's probably because of the special casing of live data in `queryJobs`, where two calls to `evaluateFilter` are done.
```
Originally added on 2014-09-24 18:49:02 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Filed the `gnt-job list --filter` bug as a separate issue 958.

This bug now continues to handle the language parsing problems mentioned in #1 and #2.
```
Originally added on 2014-09-25 13:32:02 +0000 UTC.

Mentioned Issue with GoogleCode ID 958 has been migrated to GitHub with ID #1014.
Added Labels: Component-scripts

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit 29bd36ee5872e873c0d880990b228f534cefb54c
Author: Niklas Hambuechen <niklash@google.com>
Date:   Tue Sep 30 17:02:45 2014 +0200

    query: Support legacy "=" filter also in Python
    
    We already did this for the Haskell part; this fixes the Python
    query implementation and tests.
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>


commit 8520e91269d0328f6e6d93523e1e94cf5d54a9a1
Author: Niklas Hambuechen <niklash@google.com>
Date:   Tue Sep 30 11:40:17 2014 +0200

    Query.Language: Support "!="
    
    It is specified in `man 7 ganeti`.
    
    Ideally we would drop support for it and keep only "not (==)",
    but Ganeti 2.10 supported "!=", so we keep it as supported legacy.
    
    This only affects parsing "!="; the internal representation is "not (==)".
    
    Signed-off-by: Niklas Hambuechen <niklash@google.com>
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

```
Originally added on 2014-09-30 16:37:15 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released as part of 2.14.0
```
Originally added on 2015-06-02 10:37:03 +0000 UTC.

Changed State: Released
